src/main.rs

```diff
@@ -0,0 +1,132 @@
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![feature(rustc_private, link_args)]
+extern crate getopts;
+extern crate rustc;
+extern crate rustc_driver;
+extern crate syntax;
+extern crate byteorder;
+extern crate rustc_const_math;
+extern crate mirkat;
+
+use rustc_driver::{driver, CompilerCalls};
+use rustc::ty::TyCtxt;
+use rustc::session::Session;
+use std::process;
+use rustc::hir::def_id::DefId;
+pub use rustc_const_math::ConstInt;
+
+use mirkat::machine::{Machine,Memory};
+
+/// This struct is required by the rustc API in order to implement traits which
+/// are used to hook into the compiler and customise the build pipeline. In
+/// Mirkat, we stop compilation after the MIR has been generated and then hand
+/// the MIR over to the interpreter. This means that Mirkat's invocation of rustc
+/// will never hit LLVM.
+struct MirkatCompilerCalls { }
+
+impl<'a> CompilerCalls<'a> for MirkatCompilerCalls {
+    fn build_controller(&mut self,
+                        _: &Session,
+                        _: &getopts::Matches)
+                        -> driver::CompileController<'a> {
+        let mut control = driver::CompileController::basic();
+
+        control.after_analysis.stop = rustc_driver::Compilation::Stop;
+        control.after_analysis.callback = Box::new(move |state| {
+            state.session.abort_if_errors();
+            let tcx = state.tcx.unwrap();
+            if let Some((entry_node_id, _)) = *state.session.entry_fn.borrow() {
+                let entry_def_id = tcx.hir.local_def_id(entry_node_id);
+                let _start_wrapper = tcx.lang_items().start_fn().and_then(|start_fn| {
+                    if tcx.is_mir_available(start_fn) {
+                        Some(start_fn)
+                    } else {
+                        None
+                    }
+                });
+            }
+            else {
+                panic!("No main funciton found");
+            }
+        });
+        control
+    }
+}
+
+/// Without specifying the sysroot, the rustc library can't be used.  This code
```

Comment:
Are licences compatible? Do you need to list the original license at the top of the file?

Reply:
Hmm, I don't know. That's probably a question for @ltratt. I didn't think about that.

Reply:
I've looked at Miri's license. They use the Apache 2.0 license. It contains this clause:

> The above copyright notice and this permission notice
shall be included in all copies or substantial portions
of the Software.

Do I need to include the Apache 2.0 license at the top of the `main.rs` file?

Reply:
Small snippets of code aren't copyrightable, so this is OK as-is.